src/pyclass.rs

```diff
@@ -0,0 +1,557 @@
+    fn as_ptr(&self) -> *mut ffi::PyObject {
+        (self as *const _) as *mut _
+    }
+}
+
+impl<T: PyClass> std::ops::Deref for PyClassShell<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        self.pyclass.deref()
+    }
+}
+
+impl<T: PyClass> std::ops::DerefMut for PyClassShell<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        self.pyclass.deref_mut()
+    }
+}
+
+impl<T: PyClass> ToPyObject for &PyClassShell<T> {
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        unsafe { PyObject::from_borrowed_ptr(py, self.as_ptr()) }
+    }
+}
+
+impl<T: PyClass> ToPyObject for &mut PyClassShell<T> {
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        unsafe { PyObject::from_borrowed_ptr(py, self.as_ptr()) }
+    }
+}
+
+unsafe impl<'p, T> FromPyPointer<'p> for &'p PyClassShell<T>
+where
+    T: PyClass,
+{
+    unsafe fn from_owned_ptr_or_opt(py: Python<'p>, ptr: *mut ffi::PyObject) -> Option<Self> {
+        NonNull::new(ptr).map(|p| &*(gil::register_owned(py, p).as_ptr() as *const PyClassShell<T>))
+    }
+    unsafe fn from_borrowed_ptr_or_opt(py: Python<'p>, ptr: *mut ffi::PyObject) -> Option<Self> {
+        NonNull::new(ptr)
+            .map(|p| &*(gil::register_borrowed(py, p).as_ptr() as *const PyClassShell<T>))
+    }
+}
+
+unsafe impl<'p, T> FromPyPointer<'p> for &'p mut PyClassShell<T>
+where
+    T: PyClass,
+{
+    unsafe fn from_owned_ptr_or_opt(py: Python<'p>, ptr: *mut ffi::PyObject) -> Option<Self> {
+        NonNull::new(ptr).map(|p| {
+            &mut *(gil::register_owned(py, p).as_ptr() as *const PyClassShell<T> as *mut _)
+        })
+    }
+    unsafe fn from_borrowed_ptr_or_opt(py: Python<'p>, ptr: *mut ffi::PyObject) -> Option<Self> {
+        NonNull::new(ptr).map(|p| {
+            &mut *(gil::register_borrowed(py, p).as_ptr() as *const PyClassShell<T> as *mut _)
+        })
+    }
+}
+
+/// An initializer for `PyClassShell<T>`.
+pub struct PyClassInitializer<T: PyTypeInfo> {
+    init: Option<T>,
+    super_init: Option<*mut PyClassInitializer<T::BaseType>>,
+}
+
+impl<T: PyTypeInfo> PyClassInitializer<T> {
+    pub fn from_value(value: T) -> Self {
+        PyClassInitializer {
+            init: Some(value),
+            super_init: None,
+        }
+    }
+
+    pub fn new() -> Self {
+        PyClassInitializer {
+            init: None,
+            super_init: None,
+        }
+    }
+
+    #[must_use]
+    #[doc(hiddden)]
+    pub fn init_class(self, shell: &mut T::ConcreteLayout) -> PyResult<()> {
+        macro_rules! raise_err {
+            ($name: path) => {
+                return Err(PyErr::new::<RuntimeError, _>(format!(
+                    "Base class '{}' is not initialized",
+                    $name
+                )));
+            };
+        }
+        let PyClassInitializer { init, super_init } = self;
+        if let Some(value) = init {
+            unsafe { shell.py_init(value) };
+        } else if !T::ConcreteLayout::NEED_INIT {
```

Comment:
is the condition reversed here, erroring when `NEED_INIT` is `false` seems backwards.

Reply:
> erroring when NEED_INIT is false seems backwards

Sorry I'm not sure what you are saying..., but you mean we should not raise an error for redundant initialization?

Reply:
looks good!